source/server/http/admin.cc

```diff
@@ -653,7 +663,10 @@ void AdminFilter::onComplete() {
   Buffer::OwnedImpl response;
   Http::HeaderMapPtr header_map{new Http::HeaderMapImpl};
   RELEASE_ASSERT(request_headers_);
-  Http::Code code = parent_.runCallback(path, *request_headers_, *header_map, response);
+
+  AdminStreamImpl admin_stream(*callbacks_, *request_headers_, on_destroy_callbacks_);
```

Comment:
I'm sorry, I haven't been tracking this lengthy conversation, but why doesn't AdminFilter implement AdminStream and then you just pass `*this`? That is the correct ownership model anyway IMO and avoids all of the other concerns? cc @jmarantz

Reply:
We started with that model and then changed it in response to feedback. If it makes more sense we can go back to that. @mattklein123 @jmarantz

Reply:
Alright, I don't mean to rehash a huge conversation. Personally, I think AdminFilter implementing AdminStream is the clearest model, as the filter/stream is alive for the lifetime of the admin request, so that is what I would do, but if we want to start with this we can see how it goes. Though I don't see how this can be the final solution as being stack allocated, as for handlers that continue processing they will do so outside the context of onComplete, so something will need to change?

Reply:
I don't know if it answers your question, but - In the Hystrix example, only the StreamDecoderFilterCallbacks is needed after onComplete is done with. It references to outside the AdminFilter. In Hystrix we keep pointer to it in the sink, so the pointer stay valid after the lifetime of onComplete(). The idea we agreed here is to handle this by adding methods to on_destroy_callbacks_ that will stop using the StreamDecoderFilterCallbacks pointer when onDestroy() is called. Adding this function should be done in the handler.

Reply:
I've already commented on this several times, but there is nothing happening here at the granularity of the connection. Only streams, and the filter object has the lifetime of the stream. Trust me, you want the filter to implement the interface.

Reply:
+1.  Having the filter implement the interface should work perfectly for the Hystrix case.  It will make the lifetime easier to understand, and it will incur the smallest code footprint.

Reply:
I talked about this offline with @jmarantz. I think we are all on the same page to move to the inheritance model. Sorry for the runaround @trabetti but I think important learning/exploration for all involved.